### PR TITLE
8335288: SunPKCS11 initialization will call C_GetMechanismInfo on unsupported mechanisms

### PR DESCRIPTION
**Pull request by valeriepeng**
Can someone help review this fix? Changed the required-mechanism check by checking if the particular mechanism is inside the list of enabled supported mechanisms. This should be more reliable than calling C_GetMechanismInfo(..) on the required mechanism given vendors may return various sorts of error codes.

Thanks,
Valerie

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335288](https://bugs.openjdk.org/browse/JDK-8335288): SunPKCS11 initialization will call C_GetMechanismInfo on unsupported mechanisms (**Bug** - P3)

### Reviewers
 * [Hai-May Chao](https://openjdk.org/census#hchao) (@haimaychao - Committer) Review applies to [a37ab014](https://git.openjdk.org/jdk/pull/20207/files/a37ab014c0ba5beb553ff800901b2b24d6fdfaf5)
 * [Martin Balao](https://openjdk.org/census#mbalao) (@martinuy - Committer)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20207/head:pull/20207` \
`$ git checkout pull/20207`

Update a local copy of the PR: \
`$ git checkout pull/20207` \
`$ git pull https://git.openjdk.org/jdk.git pull/20207/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20207`

View PR using the GUI difftool: \
`$ git pr show -t 20207`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20207.diff">https://git.openjdk.org/jdk/pull/20207.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20207#issuecomment-2232086241)

### CODE REVIEW

**Review by haimaychao — APPROVED** [review]

**Comment by haimaychao** [comment]
LGTM.

**Comment by valeriepeng** [comment]
> LGTM.

Thanks Haimay for the review~

**Comment by valeriepeng** [comment]
> I understand that the sample config is for a test, but are there any mechanisms we _would_ want to disable by default? It occurred to me as I was reading through the test and noticed that SHA1 was not in the disabled list for the test.

Are you asking about general PKCS11 provider configuration setting? For PKCS11 providers, users provide their provider configuration file and they decide what to disable. As for SHA-1, it's not disabled by default for SUN provider either.

**Comment by valeriepeng** [comment]
@martinuy Can you also review this? IIRC, this is introduced by the PBE PKCS12 changes earlier.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 1]
```diff
@@ -1254,6 +1254,17 @@ private void initToken(CK_SLOT_INFO slotInfo) throws PKCS11Exception {
...
         }
+
+        long[] supportedMechanisms = p11.C_GetMechanismList(slotID);
+        // start w/ the supported mechanisms, then remove the config disabled
+        // and known broken ones
+        Set<Long> enabledMechSet =
+                Arrays.stream(supportedMechanisms).boxed().collect
+                (Collectors.toCollection(HashSet::new));
+        if (config.getDisabledMechanisms() != null) {
+            enabledMechSet.removeAll(config.getDisabledMechanisms());
```
Comment (martinuy) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1265]: Should `enabledMechSet` be further restricted to `Config::enabledMechanisms` (if set)? `Config::disabledMechanisms` looks like a fallback for when `Config::enabledMechanisms` is not set, according to `Config::isEnabled`. I'd keep the logic that makes `Config::enabledMechanisms` work in pair with `Config::disabledMechanisms` in a single place, as duplicating it into two different places may lead to misalignment.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 2]
```diff
@@ -1281,10 +1292,9 @@ private void initToken(CK_SLOT_INFO slotInfo) throws PKCS11Exception {
         // the earliest entry.  When asked for "DES/CBC/PKCS5Padding", we
         // return a CKM_DES_CBC_PAD.
         final Map<Descriptor,Integer> supportedAlgs =
-                                        new HashMap<Descriptor,Integer>();
+                new HashMap<Descriptor,Integer>();
 
-        for (int i = 0; i < supportedMechanisms.length; i++) {
-            long longMech = supportedMechanisms[i];
+        for (long longMech : supportedMechanisms) {
```
Comment (martinuy) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1297]: Is the code under `if (!config.isEnabled(longMech)) {` still needed? Looks to me that we will be iterating over enabled mechanisms now.

Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1297]: Just for the record, I want to note that we will not longer be showing information about supported but disabled mechanisms here.

**Review by martinuy — CHANGES_REQUESTED** [review]
Overall, I like the change because in addition to fixing the bug we will save some `getMechanismInfo` calls for mechanisms that are supported by the token but disabled in the configuration. I made a couple of minor comments, though.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java (hunk 1 above)
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1265]: Yes, you have a good point. Let me think about it and see how to better handle this.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java (hunk 2 above)
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1297]: Right, I agree that it's better to show info about supported but disabled mechanisms.

**Review by martinuy — APPROVED** [review]
Looks good to me.

**Comment by valeriepeng** [comment]
> Looks good to me.

Thanks Martin for the review~

**Review by wangweij — APPROVED** [review]
Change looks fine.

**Comment by haimaychao** [comment]
Updated webrev looks good.

**Comment by kaarefc** [comment]
Hi. I'm the original reporter of this issue.
Thanks for the work done on this bug.
I see the the fix has been targeted at only Java 24.
Are there any plans to back port this to Java 21, which is the current LTS version, and where the regression was introduced?

**Comment by valeriepeng** [comment]
> Hi. I'm the original reporter of this issue. Thanks for the work done on this bug. I see the the fix has been targeted at only Java 24. Are there any plans to back port this to Java 21, which is the current LTS version, and where the regression was introduced?

I have just opened a request to backport this to JDK 21 update. Someone from the sustaining team should handle the backport request. Thank~

**Comment by valeriepeng** [comment]
> > Hi. I'm the original reporter of this issue. Thanks for the work done on this bug. I see the the fix has been targeted at only Java 24. Are there any plans to back port this to Java 21, which is the current LTS version, and where the regression was introduced?
> 
> I have just opened a request to backport this to JDK 21 update. Someone from the sustaining team should handle the backport request. Thank~

Correction: It'll be the OpenJDK21 maintainers to handle the backport request (not the sustaining team).

**Comment by valeriepeng** [comment]
> > > Hi. I'm the original reporter of this issue. Thanks for the work done on this bug. I see the the fix has been targeted at only Java 24. Are there any plans to back port this to Java 21, which is the current LTS version, and where the regression was introduced?
> > 
> > 
> > I have just opened a request to backport this to JDK 21 update. Someone from the sustaining team should handle the backport request. Thank~
> 
> Correction: It'll be the OpenJDK21 maintainers to handle the backport request (not the sustaining team).

@GoeLin I fixed this in jdk24 and the submitter hopes that can be backported to jdk21. Is this something that you can help with? Thanks!

**Pull request closed without merging**